File: base/src/main/vsdk/toolkit/common/statistics/RaytraceStatistics.java
package vsdk.toolkit.common;

import java.util.concurrent.atomic.LongAdder;

public final class RaytraceStatistics {
    private static final boolean ENABLED = Boolean.getBoolean("vsdk.raytrace.stats");

    private static final LongAdder primaryRays = new LongAdder();
    private static final LongAdder shadowRays = new LongAdder();
    private static final LongAdder reflectionRays = new LongAdder();
    private static final LongAdder sceneTraversals = new LongAdder();
    private static final LongAdder objectIntersectionTests = new LongAdder();
    private static final LongAdder rayWithTCalls = new LongAdder();
    private static final LongAdder rayHitInstances = new LongAdder();
    private static final LongAdder hitInfoClones = new LongAdder();
    private static final LongAdder geometryDetailComputations = new LongAdder();

    private RaytraceStatistics() {
    }

    public static boolean isEnabled() {
        return ENABLED;
    }

    public static void recordPrimaryRay() {
        if (ENABLED) {
            primaryRays.increment();
        }
    }

    public static void recordShadowRay() {
        if (ENABLED) {
            shadowRays.increment();
        }
    }

    public static void recordReflectionRay() {
        if (ENABLED) {
            reflectionRays.increment();
        }
    }

    public static void recordSceneTraversal() {
        if (ENABLED) {
            sceneTraversals.increment();
        }
    }

    public static void recordObjectIntersectionTest() {
        if (ENABLED) {
            objectIntersectionTests.increment();
        }
    }

    public static void recordRayWithT() {
        if (ENABLED) {
            rayWithTCalls.increment();
        }
    }

    public static void recordRayHitInstance() {
        if (ENABLED) {
            rayHitInstances.increment();
        }
    }

    public static void recordHitInfoClone() {
        if (ENABLED) {
            hitInfoClones.increment();
        }
    }

    public static void recordGeometryDetailComputation() {
        if (ENABLED) {
            geometryDetailComputations.increment();
        }
    }

    public static void printSummary() {
        if (!ENABLED) {
            return;
        }

        long primaryRaysSnapshot = primaryRays.sum();
        long shadowRaysSnapshot = shadowRays.sum();
        long reflectionRaysSnapshot = reflectionRays.sum();
        long sceneTraversalsSnapshot = sceneTraversals.sum();
        long objectIntersectionTestsSnapshot = objectIntersectionTests.sum();
        long rayWithTCallsSnapshot = rayWithTCalls.sum();
        long rayHitInstancesSnapshot = rayHitInstances.sum();
        long hitInfoClonesSnapshot = hitInfoClones.sum();
        long geometryDetailComputationsSnapshot = geometryDetailComputations.sum();
        long totalRays =
            primaryRaysSnapshot + shadowRaysSnapshot + reflectionRaysSnapshot;

        System.out.println("Ray statistics:");
        System.out.println("  Primary rays: " + primaryRaysSnapshot);
        System.out.println("  Shadow rays: " + shadowRaysSnapshot);
        System.out.println("  Reflection rays: " + reflectionRaysSnapshot);
        System.out.println("  Total rays cast: " + totalRays);
        System.out.println("  Scene traversals: " + sceneTraversalsSnapshot);
        System.out.println("  Object intersection tests: " + objectIntersectionTestsSnapshot);
        System.out.println("  Ray.withT calls: " + rayWithTCallsSnapshot);
        System.out.println("  RayHit instances: " + rayHitInstancesSnapshot);
        System.out.println("  Hit info clones: " + hitInfoClonesSnapshot);
        System.out.println("  Geometry detail computations: " + geometryDetailComputationsSnapshot);
        if (sceneTraversalsSnapshot > 0) {
            System.out.println(
                "  Avg. object tests / traversal: " +
                ((double) objectIntersectionTestsSnapshot / (double) sceneTraversalsSnapshot)
            );
        }
    }
}

File: typescript/base/src/main/vsdk/toolkit/common/statistics/RaytraceStatistics.ts
type StatisticsFlags = { readonly raytrace?: boolean; readonly polyhedral?: boolean };
const statisticsFlags = (globalThis as { __vitralStatistics?: StatisticsFlags }).__vitralStatistics;

/** Opt-in browser-safe counterpart of the Java system-property ray-tracing counters. */
export class RaytraceStatistics {
  private static readonly enabled = statisticsFlags?.raytrace === true;
  private static primaryRays = 0n; private static shadowRays = 0n; private static reflectionRays = 0n;
  private static sceneTraversals = 0n; private static objectIntersectionTests = 0n; private static rayWithTCalls = 0n;
  private static rayHitInstances = 0n; private static hitInfoClones = 0n; private static geometryDetailComputations = 0n;
  public static isEnabled(): boolean { return this.enabled; }
  public static recordPrimaryRay(): void { if (this.enabled) this.primaryRays++; } public static recordShadowRay(): void { if (this.enabled) this.shadowRays++; } public static recordReflectionRay(): void { if (this.enabled) this.reflectionRays++; }
  public static recordSceneTraversal(): void { if (this.enabled) this.sceneTraversals++; } public static recordObjectIntersectionTest(): void { if (this.enabled) this.objectIntersectionTests++; } public static recordRayWithT(): void { if (this.enabled) this.rayWithTCalls++; }
  public static recordRayHitInstance(): void { if (this.enabled) this.rayHitInstances++; } public static recordHitInfoClone(): void { if (this.enabled) this.hitInfoClones++; } public static recordGeometryDetailComputation(): void { if (this.enabled) this.geometryDetailComputations++; }
  public static printSummary(): void { if (!this.enabled) return; const total = this.primaryRays + this.shadowRays + this.reflectionRays; console.log(`Ray statistics:\n  Primary rays: ${this.primaryRays}\n  Shadow rays: ${this.shadowRays}\n  Reflection rays: ${this.reflectionRays}\n  Total rays cast: ${total}\n  Scene traversals: ${this.sceneTraversals}\n  Object intersection tests: ${this.objectIntersectionTests}\n  Ray.withT calls: ${this.rayWithTCalls}\n  RayHit instances: ${this.rayHitInstances}\n  Hit info clones: ${this.hitInfoClones}\n  Geometry detail computations: ${this.geometryDetailComputations}${this.sceneTraversals > 0n ? `\n  Avg. object tests / traversal: ${Number(this.objectIntersectionTests) / Number(this.sceneTraversals)}` : ""}`); }
}
